controllers/drplacementcontrol.go

```diff
@@ -1429,12 +1450,18 @@ func (d *DRPCInstance) EnsureCleanup(clusterToSkip string) error {
 		d.setDRPCCondition(&d.instance.Status.Conditions, rmn.ConditionPeerReady, d.instance.Generation,
 			metav1.ConditionFalse, rmn.ReasonCleaning, msg)
 
+		rmnutil.ReportIfNotPresent(d.reconciler.eventRecorder, d.instance, corev1.EventTypeWarning,
+			rmnutil.EventReasonCleaning, msg)
+
 		return fmt.Errorf("waiting to clean secondaries")
 	}
 
 	d.setDRPCCondition(&d.instance.Status.Conditions, rmn.ConditionPeerReady, d.instance.Generation,
 		metav1.ConditionTrue, rmn.ReasonSuccess, "Cleaned")
 
+	rmnutil.ReportIfNotPresent(d.reconciler.eventRecorder, d.instance, corev1.EventTypeNormal,
+		rmnutil.EventReasonCleanupSuccess, "Cleaned")
```

Comment:
I think the events should NOT follow 100% conditions update.  As an example, this event does not tell what "cleaned" is for. You can be verbose and say that the Peer is cleaned, but maybe not necessary. But what I think is important is to describe the events as they occur up to the initialdeployement or failover or relocate is complete. As an example for initial deployment, I would like to see a list of events similar to this:
1. Initializing deployment
2. Deploying
3. Creating VRG MW
4. Updating placementRule
5. Setting up VolSync Replication <--------------- IF VolSync is enabled
6. Cleaning up
7. Deployed

And you would also have in between failures if there are any. Example:
1. Initializing deployment
2. HomeCluster is required
...

Reply:
Adding to this,

- We should also refrain from just passing err.Error() strings as the message. As these may contain lower level messages, and would eventually land in the condition for the user to look at, the event itself should summarize the message as an failure to do something.
- If we think of documenting the events that the reconciler generates, that should be a good start for our even types and messages (IOW, which events do the cluster administrators or namespace users watch for, what are relevant to potential troubleshooting later)
- Also flooding events for a regular reconcile may not be required, i.e each state transition. We will possibly want state transitions we are waiting for (VRG to move to Primary/Seconday as an example), but not add events for all intermediate transitions. IOW think about the noise it may generate
- We would need events for major transitions, so that we have a history (e.g at some point in the past failover was started, and it never completed and a relocate was initiated etc. as 2 separate events though)

A good start here would be laying out the events that we want, and then adding them to the code.

Reply:
Thanks Benamar and Shyam. I will work on these comments. 
Need some more clarity as my initial understanding was to add events that were logically related and show progression. Now that will of course generate lot more events which may not be a good thing. We might just want major transitions or upate only. 
For example in the  failover path switchToCluster is a function which is called, inside witch we do 3 things 1, createVRGmanifest as Primary 2, Check if PVs are restored and then update userplacement rule. 
In the current code when switchToCluster returns  an error then there is an event generated (EventReasonSwitchFailed) which is generic with an error . This error might have occurred in any of the three steps that I have  mentioned above.  To my understanding None of these are major transition.
Now should I add events for each of those step as Benamar mentioned or should keep it as it is (which generic EventReasonSwitchFailed) and replace the error with a message.

Reply:
@jmishrarh It seems that we are more focused on failures more than anything. The way I look at events is basically the various states a component takes. I can look at the component's events at any point in time and know what is happening. The components we are talking about here is `DRPC` and `VRG`.  So if DRPC is waiting for PV Restore to complete, I would like to see the last event to be that. I suggest coming up with a set of events for each (DRPC/VRG), and sprinkle them out in the appropriate places.

Reply:
Thanks Benamar. I got your point. I will come up with list of events and then we can decide.